### ISSUE TITLE
current sync engine creates conflicts in CouchDB.

### ISSUE THREAD

**Issue by gr2m** [hoodiehq/hoodie#66]
hoodie.js is using CouchDB's _bulk_docs API with ?new_edits=false. I randomly generaty new _rev numbers locally and then push all changes to CouchDB. But I found out, we can't do it this way, it leads to conflicts.

the `_rev` numbers need to be properly calculated. This issue is about getting the JS version fo the algorithm and to put it into hoodie.js. 

Here is what currently happens.

``` js
// sign up for a new account or sign in to an existing account
hoodie.account.signUp('test', 'test')

// add a new test object with id 'test'
hoodie.store.add('test', { rev: '1', id: 'test' })
```

wait 2 seconds, until the POST /_api/userdb/_bulk_docs request finishes.  
http://couch.myapp.dev/userdb/test%2Ftest now looks like this:

``` json
{
    "_id": "test/test",
    "_rev": "1-owd29jdlm",
    "rev": "1",
    "createdBy": "7ddyj3l",
    "updatedAt": "2013-05-23T19:42:35.152Z",
    "createdAt": "2013-05-23T19:42:35.152Z",
    "type": "test"
}
```

now update the test object

``` js
hoodie.store.add('test', 'test', { rev: '2'})
```

Again wait to seconds, http://couch.myapp.dev/userdb/test%2Ftest now still looks the same. If you add `?conflicts=true`, you'll see:

``` json
{
    "_id": "test/test",
    "_rev": "1-owd29jdlm",
    "rev": "1",
    "createdBy": "7ddyj3l",
    "updatedAt": "2013-05-23T19:42:35.152Z",
    "createdAt": "2013-05-23T19:42:35.152Z",
    "type": "test",
    "_conflicts": [
        "1-gfzidwk7a"
    ]
}
```

That's bad. That really should not happen. We want to send a valid new revision, without a conflict being created.

If anyone could help me getting the right algorithm together to calculate the `_rev` number, that'd be great
## Update

I originally thought that the cause of the problem is that we are generating random `_rev` numbers, that is not the case though. Instead, I've found that Hoodie doesn't send the right `_revisions.ids` history in some cases, which leads to the conflicts.

**Comment by bbn** [comment]
I think before sending to the DB you have to populate the second 'test' document with the _rev of the first saved document. In other words you don't need to calculate the next _rev, you just need to pop in the _rev of the latest doc and couch will update for you.

I don't know the internals of CouchDB but it's likely the exact guessing of each _rev is something that couldn't be done locally anyway. It may use data that you don't have access to.

**Comment by gr2m** [comment]
here is the current implementation in hoodie.js:
https://github.com/hoodiehq/hoodie.js/blob/master/src/core/remote.coffee#L402

``` coffeescript
  _addRevisionTo : (attributes) ->

    try [currentRevNr, currentRevId] = attributes._rev.split /-/
    currentRevNr = parseInt(currentRevNr, 10) or 0

    newRevisionId = @_generateNewRevisionId()

    attributes._rev = "#{currentRevNr + 1}-#{newRevisionId}"
    attributes._revisions = 
      start : 1
      ids   : [newRevisionId]

    if currentRevId
      attributes._revisions.start += currentRevNr
      attributes._revisions.ids.push currentRevId
```

so what happens internally when `hoodie.store.add('test', 'test', { rev: '2'})` gets executed is the following:
- it adds a `_revisions` attribute
- it sets `_revisions.start = 2`
- it sets `_revisions.ids = ['newrandomhash', 'owd29jdlm']

This way CouchDB knows the old `_rev` id and can add the new one. So far the theory. The problem is, that `newrandomhash` cannot be random, instead the `@_generateNewRevisionId()` method has to calculate a `_rev` number based on the contents. And I don't know how

**Comment by bbn** [comment]
ok I see. I looked through the couch documentation and I understand now what you are trying to do.
shot in the dark, but do the ids in the _revisions.ids array have to be in order? the way you've got it set up now, it will be [newRevisionId, currentRevId] rather than vice-versa.

also how can this code possibly work when there is more than 1 previous revision? ie if currentRevNr is > 1.

**Comment by gr2m** [comment]
you're right, I fixed my comment, the newest id comes first in `_revisions.ids`. As long as CouchDB finds that the second id is the current rev id it should not create a conflict as it can add it to the revision tree. But I'm not an expert, I'll wait for some input by @janl or some other CouchDB dev

**Comment by dch** [comment]
why are you using ?new_edits=false at all? This is a sorta-under-the-table thing used for the replicator to allow generating conflicting versions of documents. It's documented because is is part of the core API though. If you are able to create conflicts, then you also need to be able to resolve them too -- are you sure you need this?

You're better off with normal PUT (via bulk docs if you want), optionally providing your own ids if needed, and letting couchdb handle the _rev generation. I can't imagine a situation where you want to do all of this yourself. At worst if you absolutely have to do this, at least pinch pouchdb's code.

**Comment by nnarhinen** [comment]
After talking with @dch in IRC I come to think of the following resolution:
1. use `_bulk_docs` without `new_edits=false`
2. loop through the response and check for documents with `error: 'conflict'
3. use a conflict resolution strategy

We could have two built-in strategies:
1. 'local wins': We update the revision of our document with the one from database and push our content with the new rev number
2. 'emit conflict event': We emit a conflict event that either the dev can handle automatically or allow showing some merge-view for users.

IMO conflicts have to be handled somewhere (by handling I think we can also do 'local wins' for now) but I don't see any reason to use the replicator API for pushing revisions. Also they might still end up conflicting and those conflicts have to be handled, right @dch ?

**Comment by daleharvey** [comment]
I am not familiar with how hoodie sync is implemented, but yeh this is a question of how to handle sync, not revision generation, couch is fine handling arbitrarily generated revisions. 

new_edits=false doesnt necessarily generate conflicts, it just requires that you send it parent information as the _revisions field of a document, that code is https://github.com/daleharvey/pouchdb/blob/master/src/pouch.utils.js#L110 in pouch (its probably the least readable code, needs a bit of a refactor)

You need to decide how to handle conflicts, personally I would prefer using new_edits=false but the only difference that causes is when you handle the conflicts, at write time or any time after the write

**Comment by daleharvey** [comment]
and just to mention the revision generation, its 'possible' to generate revisions in the same way as couchdb, but I dont think it is worth the effort (the revision is calculated from the erlang internal representation of the data, its not guaranteed by erlang and a leaky abstraction imo), the only effect it causes is that in couchdb if you make the exact same changes to a document on 2 servers and replicate between them, it wont generate conflicts whereas randomly generated revisions will

**Comment by gr2m** [comment]
Thanks guys for all your help! I see things way more clear now!

So hoodie creates conflicts, but after looking deeper into it,  the reason is that in some cases it doesn't send the right `_revisions.ids` history. So it must be a bug in Hoodie itself, generating the IDs locally is not a problem as I thought initially. The only benefit from calculating the `_rev` values is if the same changes are made on two devices at the same time, as it won't lead to conflicts, that I understand now.

Here's the reason why I want to use the `new_edits=false` flag:
1. I want to make sure, that no user data gets lost. 
2. I want conflict resolution to happen on our servers, and only leave edge cases to the frontend.

My experience is that conflicts rarely happen. And if they happen, it's often times because a user started working on device A, but it didn't sync for whatever reason. Then continues on device B. A few days later, he starts the app on device A again, and cannot sync because there is a conflict. I've put automated conflict resolutions strategies in place, but still at some edge cases, some kind of race kondition appears and it kills the browser. I'd rather make the implementation in the browser a simple & stupid as possible, and keep more of the complexity on the server side, as this is where we can debug and react if things go funky.

It's very hard to debug browsers, especially when data is stored locally. I try to avoid that. And the good thing about Hoodie is that the architecture is made for a "conflict resolution worker", that listens to all changes and reacts on it.

On the other side you're the CouchDB experts, if you all agree we should do the sync without the `new_edits=false`, we should listen to you I guess? I'll see what @janl thinks. 

Thanks again for your valuable input

**Comment by janl** [comment]
> Thanks guys for all your help! I see things way more clear now!
> 
> So hoodie creates conflicts, but after looking deeper into it, the reason is that in some cases it doesn't send the right _revisions.ids history.

Correct, thanks to @daleharvey for clearing this up :)

> So it must be a bug in Hoodie itself, generating the IDs locally is not a problem as I thought initially.

Yeah, sorry for putting you on the wrong track there initially!

> The only benefit from calculating the _rev values is if the same changes are made on two devices at the same time, as it won't lead to conflicts, that I understand now.
> 
> Here's the reason why I want to use the new_edits=false flag:
> 
>   • I want to make sure, that no user data gets lost.
>   • I want conflict resolution to happen on our servers,

These are both very good reasons.

When @dch said "This is a sorta-under-the-table thing…" that is not entirely accurate, it is part of the public API of CouchDB and you should totally use it, if you are writing a thing that behaves like a replicator, which is what Hoodie sync is.

tl;dr: we are on the right track here :)

**Comment by gr2m** [comment]
here's @janl's comment in readable. Don't know why his comment is broken 

> Thanks guys for all your help! I see things way more clear now!
> 
> So hoodie creates conflicts, but after looking deeper into it, the reason is that in some cases it doesn't send the right _revisions.ids history.

Correct, thanks to @daleharvey for clearing this up :)

> So it must be a bug in Hoodie itself, generating the IDs locally is not a problem as I thought initially.

Yeah, sorry for putting you on the wrong track there initially!

> The only benefit from calculating the _rev values is if the same changes are made on two devices at the same time, as it won't lead to conflicts, that I understand now.
> 
> Here's the reason why I want to use the new_edits=false flag:
> 
>   • I want to make sure, that no user data gets lost.
>   • I want conflict resolution to happen on our servers,

These are both very good reasons.

When @dch said "This is a sorta-under-the-table thing…" that is not entirely accurate, it is part of the public API of CouchDB and you should totally use it, if you are writing a thing that behaves like a replicator, which is what Hoodie sync is.

tl;dr: we are on the right track here :)

**Comment by dch** [comment]
On 25 May 2013 16:17, Jan Lehnardt notifications@github.com wrote:

> On May 24, 2013, at 13:01 , Gregor Martynus notifications@github.com
> wrote:
> 
> > Here's the reason why I want to use the new_edits=false flag:
> > 
> > • I want to make sure, that no user data gets lost.
> > • I want conflict resolution to happen on our servers,
> 
> These are both very good reasons.
> 
> When @dch said "This is a sorta-under-the-table thing…" that is not
> entirely accurate, it is part of the public API of CouchDB and you should
> totally use it, if you are writing a thing that behaves like a replicator,
> which is what Hoodie sync is.
> 
> This was my primary concern - if you are using hoodie as a replicator
> itself then this makes total sense. Looking forwards to this!

A+
Dave

**Comment by gr2m** [comment]
cool, thanks for the feedback, we'll leave it this way then. 

I fixed a few issues and made some tests, looks good now, going to close this one, cheers!
